networkx/algorithms/flow/networksimplex.py

```diff
@@ -180,43 +495,23 @@ def network_simplex(G, demand="demand", capacity="capacity", weight="weight"):
     if len(G) == 0:
         raise nx.NetworkXError("graph has no nodes")
 
-    # Number all nodes and edges and hereafter reference them using ONLY their
-    # numbers
-
-    N = list(G)  # nodes
-    I = {u: i for i, u in enumerate(N)}  # node indices
-    D = [G.nodes[u].get(demand, 0) for u in N]  # node demands
-
-    inf = float("inf")
-    for p, b in zip(N, D):
-        if abs(b) == inf:
-            raise nx.NetworkXError(f"node {p!r} has infinite demand")
-
     multigraph = G.is_multigraph()
-    S = []  # edge sources
-    T = []  # edge targets
-    if multigraph:
-        K = []  # edge keys
-    E = {}  # edge indices
-    U = []  # edge capacities
-    C = []  # edge weights
 
-    if not multigraph:
-        edges = G.edges(data=True)
-    else:
-        edges = G.edges(data=True, keys=True)
-    edges = (e for e in edges if e[0] != e[1] and e[-1].get(capacity, inf) != 0)
-    for i, e in enumerate(edges):
-        S.append(I[e[0]])
-        T.append(I[e[1]])
-        if multigraph:
-            K.append(e[2])
-        E[e[:-1]] = i
-        U.append(e[-1].get(capacity, inf))
-        C.append(e[-1].get(weight, 0))
+    # extracting data essential to problem
+    DEAF = _DataEssentialsAndFunctions(
+        G, multigraph, demand=demand, capacity=capacity, weight=weight
+    )
 
-    for e, c in zip(E, C):
-        if abs(c) == inf:
+    ###########################################################################
+    # Quick Error Detection
+    ###########################################################################
```

Comment:
Could the error/infeasibility detection + this initial setup be moved inside the `_DataEssentialsAndFunctions` constructor? Was there a specific reason why the error checking was left outside of the class? If not, and it's possible to move all of this checking/initialization inside the constructor/setup methods, then it might be possible to simplify the `_DataEssentialsAndFunctions` signature. For example, it'd be nice if `multigraph = G.is_multigraph()` were only necessary inside the class, eliminating the need to pass around the `multigraph` parameter.

Reply:
I thought of `_DataEssentialsAndFunctions` as a way of abstracting away the data and related functions. Hence, I did not consider the possibility of including checks within it. But we can certainly include initial checks in the `_DataEssentialsAndFunctions` constructor. But then we would have to choose between letting the infeasibility and unboundedness checks ( they need to be after pivot loop) be in `network_simplex` and other checks in `_DataEssentialsAndFunctions`, which seems to be a very scattered way of managing checks (or is it a better way? if so, we can include it there), or we can have methods in `_DataEssentialsAndFunctions` for all of these checks (or is there any other better way?). And about the  Initialization part, it would be best to include it in the `initialize_spanning_tree` call which would make that part of `network_simplex` clean. the problem with `multigraph` is that it is also required in flow cost calculation but we can still get rid of the multigraph parameter by `self.multigraph = G.is_multigraph()` in `_DataEssentialsAndFunctions` constructor and using `DEAF.multigraph` instead of `multigraph` in `network_simplex`.

Reply:
> But then we would have to choose between letting the infeasibility and unboundedness checks ( they need to be after pivot loop) be in network_simplex and other checks in _DataEssentialsAndFunctions, which seems to be a very scattered way of managing checks (or is it a better way? if so, we can include it there), or we can have methods in _DataEssentialsAndFunctions for all of these checks (or is there any other better way?)

Thanks for the explaining the thought process here - I was mostly curious if there were a *simple* way of reducing the number of parameters in the constructor, but it's a minor point and not worth spending cycles on IMO. I think the organization you've chosen is entirely fine and would vote to leave it as is :)

> And about the Initialization part, it would be best to include it in the initialize_spanning_tree call which would make that part of network_simplex clean.

I think I agree - it'd be great to get rid of the pre-initialization-initializations (i.e. defining the class attributes and setting them to `None`) and I definitely would like to get rid of the `_spanning_tree_initialized` flag, which seems like the way this is headed.

> the problem with multigraph is that it is also required in flow cost calculation but we can still get rid of the multigraph parameter by self.multigraph = G.is_multigraph() in _DataEssentialsAndFunctions constructor and using DEAF.multigraph instead of multigraph in network_simplex.

Yeah, I thought it'd be nice to get rid of this extra parameter since it is always available via the `G.is_multigraph` method, but I think it's fine for now - no sense in reorganizing the entire code structure for such a minor consideration :).